File: tests/test_model.R
#' Test that the HIV Kenya model runs successfully.
#'
#' Run from repo root:
#'   Rscript tests/test_model.R
#' Or with testthat:
#'   Rscript -e "testthat::test_file('tests/test_model.R')"

library(testthat)

# Source the model; if running from tests/, step up to repo root first
if (!file.exists("hiv_model.R")) setwd("..")
source("hiv_model.R")


test_that("HIV model runs and produces valid results", {

  # Create and run the sim
  sim <- make_sim(n_agents = 1000L)
  sim$run()

  # Extract results
  res  <- sim$results$hiv
  prev <- as.numeric(res$prevalence$values)
  art  <- as.numeric(res$n_on_art$values)

  # Check prevalence
  expect_true(all(prev > 0), label = "Expect nonzero prevalence at all timepoints")
  expect_gt(prev[length(prev)], prev[1], label = "Expect prevalence to increase during the sim")

  # Check ART
  expect_equal(art[1], 0, label = "Expect no one on ART at simulation start")
  expect_gt(art[length(art)], 0, label = "Expect people on ART at simulation end")
})

File: tests/test_model.py
"""
Test that the HIV Kenya model runs successfully.
"""

import sys; sys.path.insert(0, '..') # Add parent folder to get hiv_model
import numpy as np
import sciris as sc
from hiv_model import make_sim


def test_hiv_model(do_plot=False):
    """Test that the model can be created and run without errors."""
    
    # Create and run the sim
    sim = make_sim(n_agents=1000)
    sim.run()
    
    # Do simple checks
    res = sim.results.hiv
    prev = res.prevalence
    art = res.n_on_art
    assert np.all(prev > 0), 'Expect nonzero prevalence at all timepoints'
    assert prev[-1] > prev[0], 'Expect prevalence to increase during the sim'
    assert art[0] == 0, 'Expect no one on ART at simulation start'
    assert art[-1] > 0, 'Expect people on ART at simulation end'
    
    if do_plot:
        sim.plot('hiv_prevalence_15_49')
    
    return sim


if __name__ == '__main__':
    T = sc.timer()
    do_plot = True
    
    sim = test_hiv_model(do_plot=do_plot)
    
    T.toc()
